cores/esp8266/Esp.cpp

```diff
@@ -720,195 +761,127 @@ static SpiFlashOpResult spi_flash_write_puya(uint32_t offset, uint32_t *data, si
+    auto flash_write = [&](uint32_t address, uint8_t* data, size_t size) {
+        return flashWrite(address, reinterpret_cast<uint32_t*>(data), size);
+    };
+
+    auto flash_read = [](uint32_t address, uint8_t* data, size_t size) {
+        return spi_flash_read(address, reinterpret_cast<uint32_t*>(data), size) == SPI_FLASH_RESULT_OK;
+    };
+
+    alignas(alignof(uint32_t)) uint8_t buf[FLASH_PAGE_SIZE];
 
-    if (pageBreak) {
-        size_t byteCount = 4 - (address % 4);
+    size_t written = 0;
 
-        if (!flashReplaceBlock(address, data, byteCount)) {
+    if (!isAlignedAddress(address)) {
+        auto r_addr = alignBeforeAddress(address);
+        auto c_off = address - r_addr;
+        auto wbytes = Alignment - c_off;
+
+        wbytes = std::min(wbytes, size);
+
+        if (spi_flash_read(r_addr, reinterpret_cast<uint32_t*>(&buf[0]), Alignment) != SPI_FLASH_RESULT_OK) {
             return 0;
         }
-        // We will now have aligned address, so we can cross page boundaries
-        currentOffset += byteCount;
-        // Realign size to 4
-        sizeLeft = (size - byteCount) & ~3;
-    }
 
-    while (sizeLeft) {
-        size_t willCopy = std::min(sizeLeft, sizeof(alignedData));
-        memcpy(alignedData, data + currentOffset, willCopy);
-        // We now have address, data and size aligned to 4 bytes, so we can use aligned write
-        if (!flashWrite(address + currentOffset, alignedData, willCopy))
-        {
-            return 0;
+#if PUYA_SUPPORT
+        if (getFlashChipVendorId() == SPI_FLASH_VENDOR_PUYA) {
+            for (size_t i = 0; i < wbytes ; ++i) {
+                buf[c_off + i] &= data[i];
+            }
+        } else {
+#endif
+            memcpy(&buf[c_off], data, wbytes);
+#if PUYA_SUPPORT
         }
-        sizeLeft -= willCopy;
-        currentOffset += willCopy;
-    }
+#endif
 
-    return currentOffset;
-}
+        if (spi_flash_write(r_addr, reinterpret_cast<uint32_t*>(&buf[0]), Alignment) != SPI_FLASH_RESULT_OK) {
+            return wbytes;
+        }
 
-bool EspClass::flashWritePageBreak(uint32_t address, const uint8_t *data, size_t size) {
-    if (size > 4) {
-        return false;
-    }
-    size_t pageLeft = FLASH_PAGE_SIZE - (address % FLASH_PAGE_SIZE);
-    size_t offset = 0;
-    size_t sizeLeft = size;
-    if (pageLeft > 3) {
-        return false;
+        address += wbytes;
+        data += wbytes;
+        written += wbytes;
+        size -= wbytes;
     }
 
-    if (!flashReplaceBlock(address, data, pageLeft)) {
-        return false;
-    }
-    offset += pageLeft;
-    sizeLeft -= pageLeft;
-    // We replaced last 4-byte block of the page, now we write the remainder in next page
-    if (!flashReplaceBlock(address + offset, data + offset, sizeLeft)) {
-        return false;
+    while (size > 0) {
+        auto len = std::min(size, sizeof(buf));
```

Comment:
Maybe have a double check on what is returned by `sizeof(buf)` here?
I always find it tricky to see for myself whether the `sizeof` is returning the size of the array or the size of a pointer.
In this instance, it should be `FLASH_PAGE_SIZE`

Reply:
No point guessing when IDE / language server in any editor will display this b/c it's a compile time value :)
But I suppose this could be `std::size(buf);` instead

Reply:
Well my "code review IDE" (also called "brain") is lacking this feature. ;)  It is only highlighting code which looks like code I have spent numerous hours on to fix in the past where I missed such a mistake.
Apart from that you still need to hover your mouse on those parts of the code in an IDE that does show you what it will do and that was my suggestion, to actually do that ;)

Reply:
I did just now test ESPEasy with the code changes of this PR.
It does seem to work fine.

Reply:
Please do.